lib/luanetfilter.c

```diff
@@ -24,26 +24,136 @@
 #define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
 #include <linux/module.h>
 #include <linux/version.h>
+#include <linux/types.h>
 #include <linux/netfilter.h>
 #include <linux/netfilter/x_tables.h>
 
 #include <lua.h>
 #include <lauxlib.h>
 #include <lunatik.h>
 
+typedef enum luanetfilter_type_e {
+	luanetfilter_hook_match,
+	luanetfilter_hook_target,
+	luanetfilter_hook_nfhook,
+} luanetfilter_type_t;
+
 typedef struct luanetfilter_xtable_s {
 	lunatik_object_t *runtime;
+	union {
+		struct xt_match *match;
+		struct xt_target *target;
+		struct nf_hook_ops *nfhook;
+	} hook;
+	luanetfilter_type_t type;
 } luanetfilter_xtable_t;
 
-static int luanetfilter_xtable(lua_State *L);
+static void luanetfilter_release(void *private);
+static int luanetfilter_xtable_match(lua_State *L);
+static int luanetfilter_xtable_target(lua_State *L);
+
+/* Verifies the type of "opt" in the table at "idx" and returns true if the value can be extracted.
+ * If the value is not required, it returns false.
+ * If the value is required and the type is invalid, it raises an error.
+ *
+ * TODO: create a lunatik_checkfield() API to check the type of a table field and
+ * 	raise a proper error message
+ * */
+static bool luanetfilter_xtable_opt(lua_State *L, int idx, const char *opt, int type, bool req)
+{
+	lua_getfield(L, idx, opt);
+	int type_ = lua_type(L, -1);
+	if (type_ != type) {
+		// LUA_TNIL is 0 and LUA_TNONE is -1
+		if (type_ <= 0 && !req) {
+			/* the value was not required hence, should not try to extract it */
+			return false;
+		}
+		luaL_error(L, "invalid type for %s, expected %s", opt, lua_typename(L, type));
+	}
+
+	/* true if the value can be extracted (for optional parameters) */
+	return true;
+};
 
-static void luanetfilter_release(void *private)
+#define LUANETFILTER_XTABLE_REG(mode)						\
+static int luanetfilter_xtable_##mode(lua_State *L) {				\
+	lunatik_object_t *object;						\
+	luanetfilter_xtable_t *luaxtable;					\
+										\
+	luaL_checktype(L, 1, LUA_TTABLE);					\
+										\
+	luanetfilter_xtable_opt(L, 1, "name", LUA_TSTRING, true);		\
+	const char *name = lua_tostring(L, -1);					\
+										\
+	luanetfilter_xtable_opt(L, 1, "revision", LUA_TNUMBER, true);		\
+	uint8_t revision = (uint8_t)lua_tointeger(L, -1);			\
+										\
+	luanetfilter_xtable_opt(L, 1, "family", LUA_TNUMBER, true);		\
+	unsigned short family = (unsigned short)lua_tointeger(L, -1);		\
+										\
+	unsigned proto = luanetfilter_xtable_opt(L, 1, "proto", LUA_TNUMBER, false) ? \
+			 (unsigned short)lua_tointeger(L, -1) : 0;		\
+										\
+	luanetfilter_xtable_opt(L, 1, "checkentry", LUA_TFUNCTION, true);	\
+	luanetfilter_xtable_opt(L, 1, "destroy", LUA_TFUNCTION, true);		\
+	luanetfilter_xtable_opt(L, 1, #mode, LUA_TFUNCTION, true);		\
```

Comment:
I would move the code that doesn't depend on `##mode` to an auxiliary function.. I think it will clean up the code a bit and make it more readable..